### PR TITLE
Issue #2855119 by mglaman, niko-: Use a separate order processor for …

### PR DESCRIPTION
**Pull request by mglaman**
…coupon based promotions

https://www.drupal.org/node/2855119

### CODE REVIEW

File: modules/promotion/src/PromotionOrderProcessor.php [hunk 1]
```diff
@@ -43,6 +43,11 @@ public function process(OrderInterface $order) {
     $order_type = $this->orderTypeStorage->load($order->bundle());
     $promotions = $this->promotionStorage->loadValid($order_type, $order->getStore());
     foreach ($promotions as $promotion) {
+      // Coupons have a separate order processor.
+      if (!$promotion->get('coupons')->isEmpty()) {
```
Comment (bojanz) [inline modules/promotion/src/PromotionOrderProcessor.php]: Shouldn't we instead ensure that $this->promotionStorage->loadValid() doesn't return promotions that have coupons? We can filter them out at the query level, skipping the loading of potentially dozens of entities.

**Comment by mglaman** [comment]
@bojanz yeah I was wondering if we should add a flag of `$with_coupons = FALSE` to the method. But I suppose restricting anything with coupons makes sense, because nothing with a coupon would be valid unless the coupon was applied. I'll add a `noexists` check.

**Comment by mglaman** [comment]
Also, let's not merge this until #646 is in, because it copies its code, but missing update.

**Comment by mglaman** [comment]
This failure is a test only issue :/ manually running proves it works.

File: modules/promotion/src/CouponOrderProcessor.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected $orderTypeStorage;
+
+  /**
+   * The coupon storage.
+   *
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected $promotionCouponStorage;
```
Comment (bojanz) [inline modules/promotion/src/CouponOrderProcessor.php]: Why not just call it $couponStorage? This feels too verbose

File: modules/promotion/src/CouponOrderProcessor.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace Drupal\commerce_promotion;
+
+use Drupal\commerce_order\Entity\OrderInterface;
+use Drupal\commerce_order\OrderProcessorInterface;
+use Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\PromotionOfferInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+
+/**
+ * Applies promotions to orders during the order refresh process.
```
Comment (bojanz) [inline modules/promotion/src/CouponOrderProcessor.php]: Wrong docblock. Let's do "Applies coupon promotions to orders during the order refresh process." and a @see to the PromotionOrderProcessor. We also want PromotionOrderProcessor to have a @see to this one.

File: modules/promotion/src/CouponOrderProcessor.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  /**
+   * The coupon storage.
+   *
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected $promotionCouponStorage;
+
+  /**
+   * Constructs a new PromotionOrderProcessor object.
```
Comment (bojanz) [inline modules/promotion/src/CouponOrderProcessor.php]: Wrong class name.

File: modules/promotion/src/CouponOrderProcessor.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      return;
+    }
+
+    $order_type = $this->orderTypeStorage->load($order->bundle());
+    /** @var \Drupal\commerce_promotion\Entity\CouponInterface[] $coupons */
+    $coupons = $order->get('coupons')->referencedEntities();
+    foreach ($coupons as $index => $coupon) {
+      /** @var \Drupal\commerce_promotion\Entity\PromotionInterface $promotion */
+      $promotion = $this->promotionStorage->loadByCoupon($order_type, $order->getStore(), $coupon);
+      if (!$promotion) {
```
Comment (bojanz) [inline modules/promotion/src/CouponOrderProcessor.php]: Needs comment to explain why $promotion could be FALSE.

File: modules/promotion/src/CouponOrderProcessor.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        $order->get('coupons')->removeItem($index);
+        continue;
+      }
+
+      /** @var \Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\PromotionOfferInterface $plugin */
+      $plugin = $promotion->get('offer')->first()->getTargetInstance();
```
Comment (bojanz) [inline modules/promotion/src/CouponOrderProcessor.php:73]: Isn't there a getter to replace $promotion->get('offer')->first()->getTargetInstance()?

File: modules/promotion/src/PromotionStorage.php [hunk 7]
```diff
@@ -42,7 +42,10 @@ protected function loadValidQuery(OrderTypeInterface $order_type, StoreInterface
    * {@inheritdoc}
    */
   public function loadValid(OrderTypeInterface $order_type, StoreInterface $store) {
-    $query = $this->loadValidQuery($order_type, $store);
+    $query = $this->loadQuery($order_type, $store);
+    // We do not want promotions with coupons to be loaded in this query,
```
Comment (bojanz) [inline modules/promotion/src/PromotionStorage.php]: I don't understand this comment. Perhaps something like "Only load promotions without coupons. Promotions with coupons are loaded coupon-first in a different process." ?

File: modules/promotion/src/PromotionStorage.php [hunk 8]
```diff
@@ -23,7 +23,7 @@ class PromotionStorage extends CommerceContentEntityStorage implements Promotion
    * @return \Drupal\Core\Entity\Query\QueryInterface
    *   The entity query.
    */
-  protected function loadValidQuery(OrderTypeInterface $order_type, StoreInterface $store) {
+  protected function loadQuery(OrderTypeInterface $order_type, StoreInterface $store) {
```
Comment (bojanz) [inline modules/promotion/src/PromotionStorage.php]: While we're at it, protected methods should go after the public ones.

**Review by bojanz — CHANGES_REQUESTED** [review]

File: modules/promotion/src/CouponOrderProcessor.php (hunk 6 above)
Comment (mglaman) [reply modules/promotion/src/CouponOrderProcessor.php:73]: Nope. We should add a `getOfferPlugin` to interface.

**Pull request merged**